shell_integration/MacOSX/OwnCloudFinderSync/FinderSyncExt/FinderSync.m

```diff
@@ -101,54 +115,43 @@ - (NSMenu *)menuForMenuKind:(FIMenuKind)whichMenu
 		}
 	}];
 
+	NSString *paths = [self selectedPathsSeparatedByRecordSeparator];
+	// calling this IPC calls us back from client with several MENU_ITEM entries and then our askOnSocket returns again
+	[_syncClientProxy askOnSocket:paths query:@"GET_MENU_ITEMS"];
```

Comment:
I'm sorry, but I don't understand how it's guaranteed that the `END` has arrived before `askOnSocket` returns. I could imaging it waiting for a single message to be received, but how does it know to block until all these lines have arrived?

(I see that we have a custom `socketapisocket_mac.mm` and I don't know the details of how our OSX communication works - so if you're cetain - that's fine with me. To verify, maybe add a `sleep(5)` after sending the `BEGIN` in `GET_MENU_ITEMS` and check whether that breaks it?)

Reply:
I don't know either.
I'm suspect the only way this can work now is because Qt uses the event loop to flush the "anwser" on the socket.
Would be nice if this was propertly understood though.

Reply:
It's direct synchronous IPC. 
https://gist.github.com/guruz/a49cf1497844599bba58f445216c41cf#file-gistfile1-txt-L23 <- the Extension asks for the menu entries
https://gist.github.com/guruz/a49cf1497844599bba58f445216c41cf#file-gistfile1-txt-L2 <- the extension receives the 4 menu entries (and the BEGIN and END)

Reply:
@guruz Okay, so the extension synchronously calls into the client and that synchronously calls the receiving function in the extension - and the `askOnSocket` only returns when the client function has returned? That would work.

Reply:
(It's not a socket on the macOS implementation, even if we call it SocketApi.)